### ISSUE TITLE
ChromatiCraft progress overlay is covered by minimap

### ISSUE THREAD

**Issue by CrystalMageX** [ReikaKalseki/Reika_Mods_Issues#791]
ChromatiCraft [renders](https://github.com/ReikaKalseki/ChromatiCraft/blob/1d655cba8cace1a8b72f1dddefc6acd8af1f28dd/Auxiliary/ChromaOverlays.java#L418) its progress overlays in the top right corner. Most players use minimaps, and they are by default, or people like them, in the top right corner as well. They cover the overlays, and make them impossible to read, if not unable to been seen at all.
- A **configuration option** should be added to put the notifications on the left instead.
- It is [being rendered](https://github.com/ReikaKalseki/ChromatiCraft/blob/1d655cba8cace1a8b72f1dddefc6acd8af1f28dd/Auxiliary/ChromaOverlays.java#L141) in the `RenderGameOverlayEvent.Pre` event. Notifications are a **foreground**, like achievements, _not_ a **background** image, like the pylon washout. They should be rendered in `RenderGameOverlayEvent.Post`, with a above normal priority, to appear above [at least one](https://github.com/daveyliam/mapwriter/blob/20a16a32adf4e9bcd5c984043b18a73cbaf84ef0/src/main/java/mapwriter/forge/MwForge.java#L53) minimap mod.

**Comment by ReikaKalseki** [comment]
It is supposed to render on top, and I have some GL hacks - that work for VoxelMap (which is the only minimap I care to use) at least - to accomplish that. Namely rendering at a high z-level and thus overriding render order. If it does not work for another map, there may not be much that can be done, since I render just under the near clipping plane.

Location configs are not feasible. For one, the progress notification rendering is location-sensitive (think the folding logic). On top of that, the top-left area is nominally occupied by the player element buffer display. Also, even were I to add a config like the element buffer, I cannot add full control over location, only corner, which I know from experience people just ignore and then go on to ask for full locational control - often even asking for a drag-and-drop ingame functionality with which to do so.

**Comment by TomeWyrm** [comment]
Move your minimap if it bugs you. I just put up with it, I rarely need my minimap during a progression unlock. Honestly I wish there was an option to see the server progress broadcast messages in single player and hide the overlay altogether. I miss the progression "toast" more often than not.

**Comment by Barhandar** [comment]
Covered by Mapwriter minimap. Neither VoxelMap (speaking of which, Reika, do you use the one with large map included? If so, how? It crashes for me on newer packs.) nor Journeymap obscure the progress overlay. In fact, Journeymap's minimap displays BELOW VoxelMap's!

**Comment by ReikaKalseki** [comment]
> Honestly I wish there was an option to see the server progress broadcast messages in single player and hide the overlay altogether.

Save the removal of the overlays, that _is_ an option in v13.

> (speaking of which, Reika, do you use the one with large map included? If so, how? It crashes for me on newer packs.)

Yes. I have never had an issue; I merely installed it (and its dependency Liteloader).

> Covered by Mapviewer minimap.

Is its source available?

**Comment by TomeWyrm** [comment]
Wait, there is something making the display of a chat message difficult in single player? Because I like knowing I progressed, I just always miss the overlay

**Comment by ReikaKalseki** [comment]
?

**Comment by Barhandar** [comment]
> Is its source available?

Yeah, [though on Bitbucket](https://bitbucket.org/ProfMobius/mapwriter_opis/src). I've made a mistake in name - it's writer, not viewer.

**Comment by TomeWyrm** [comment]
Your comment was confusing. It lead me to believe that there is no option to see the broadcast messages in single player. Disabling the overlay without presenting that information somewhere else seems to run counter to your usual style.

**Comment by ReikaKalseki** [comment]
I found [this](https://bitbucket.org/ProfMobius/mapwriter_opis/src/3f2e791e838543d87f5effa9234a524d1ca48b85/mapwriter/map/MapRenderer.java?at=default&fileviewer=file-view-default#MapRenderer.java-243):

```
// translate to center of minimap
    // z is -2000 so that it is drawn above the 3D world, but below minecraft
    // GUI screens
    GL11.glTranslated(this.mapMode.xTranslation, this.mapMode.yTranslation, -2000.0);
```

Mine _should_ work with that...

**Comment by CrystalMageX** [comment]
This seems to also affect the [version checker popups](https://github.com/ReikaKalseki/DragonAPI/blob/e7f465a8a273f80d4c25496d1a6b0755145ee2d5/Auxiliary/PopupWriter.java#L60) as well.

> Namely rendering at a high z-level and thus overriding render order.

2D rendering usually disables the depth test to avoid z-fighting, ignoring the z-level. If it's not disabled already for all GUI rendering, a minimap mod could disable it to avoid problems drawing the many layers of the minimap.

**Comment by ReikaKalseki** [comment]
Can I forcibly re-enable it? What is the flag?

**Comment by CrystalMageX** [comment]
You can't re-enable it unless you use ASM to patch other mods' rendering. Imagine rendering as artists painting on a canvas.

Think of the mods being the different artists, the render order be the order that they paint, and the z-level be the shininess of the paint. When `GL11.GL_DEPTH_TEST` is turned _on_, and the artist is about to paint over someone else's work that is shinier, they say "That thing looks more important than my work. It deserves respect, and my art is not worthy enough for the space it takes," and draw around it. When the depth test is turned _off_, the artist, only caring that they have gone later, says "I don't care about this masterpiece I'm painting over and going to destroy, I'm just going to do my own work," and paints over it.

In **OpenGL**, any code can change the screen contents to **whatever they want**. A mod could even clear the screen. You can, however, draw later and draw on top of _them_. (or make ASM patches for every single minimap)

**Comment by ReikaKalseki** [comment]
Yes, if they are disabling Depth Test, there is nothing I can do to fix it.

**Comment by CrystalMageX** [comment]
Just move all of the overlays to `Post`.

**Comment by ReikaKalseki** [comment]
That will break all the others.
